File: nicegui/elements/codemirror.py
import bisect
from itertools import accumulate, zip_longest
from pathlib import Path
from typing import List, Literal, Optional, Tuple, cast, get_args

from nicegui.elements.mixins.disableable_element import DisableableElement
from nicegui.elements.mixins.value_element import ValueElement
from nicegui.events import GenericEventArguments, Handler, ValueChangeEventArguments

SUPPORTED_LANGUAGES = Literal[
    'Angular Template',
    'APL',
    'ASN.1',
    'Asterisk',
    'Brainfuck',
    'C',
    'C#',
    'C++',
    'Clojure',
    'ClojureScript',
    'Closure Stylesheets (GSS)',
    'CMake',
    'Cobol',
    'CoffeeScript',
    'Common Lisp',
    'CQL',
    'Crystal',
    'CSS',
    'Cypher',
    'Cython',
    'D',
    'Dart',
    'diff',
    'Dockerfile',
    'DTD',
    'Dylan',
    'EBNF',
    'ECL',
    'edn',
    'Eiffel',
    'Elm',
    'Erlang',
    'Esper',
    'F#',
    'Factor',
    'FCL',
    'Forth',
    'Fortran',
    'Gas',
    'Gherkin',
    'Go',
    'Groovy',
    'Haskell',
    'Haxe',
    'HTML',
    'HTTP',
    'HXML',
    'IDL',
    'Java',
    'JavaScript',
    'Jinja2',
    'JSON',
    'JSON-LD',
    'JSX',
    'Julia',
    'Kotlin',
    'LaTeX',
    'LESS',
    'Liquid',
    'LiveScript',
    'Lua',
    'MariaDB SQL',
    'Markdown',
    'Mathematica',
    'Mbox',
    'mIRC',
    'Modelica',
    'MS SQL',
    'MscGen',
    'MsGenny',
    'MUMPS',
    'MySQL',
    'Nginx',
    'NSIS',
    'NTriples',
    'Objective-C',
    'Objective-C++',
    'OCaml',
    'Octave',
    'Oz',
    'Pascal',
    'Perl',
    'PGP',
    'PHP',
    'Pig',
    'PLSQL',
    'PostgreSQL',
    'PowerShell',
    'Properties files',
    'ProtoBuf',
    'Pug',
    'Puppet',
    'Python',
    'Q',
    'R',
    'RPM Changes',
    'RPM Spec',
    'Ruby',
    'Rust',
    'SAS',
    'Sass',
    'Scala',
    'Scheme',
    'SCSS',
    'Shell',
    'Sieve',
    'Smalltalk',
    'SML',
    'Solr',
    'SPARQL',
    'Spreadsheet',
    'SQL',
    'SQLite',
    'Squirrel',
    'sTeX',
    'Stylus',
    'Swift',
    'SystemVerilog',
    'Tcl',
    'Textile',
    'TiddlyWiki',
    'Tiki wiki',
    'TOML',
    'Troff',
    'TSX',
    'TTCN',
    'TTCN_CFG',
    'Turtle',
    'TypeScript',
    'VB.NET',
    'VBScript',
    'Velocity',
    'Verilog',
    'VHDL',
    'Vue',
    'Web IDL',
    'WebAssembly',
    'XML',
    'XQuery',
    'Xù',
    'Yacas',
    'YAML',
    'Z80',
]


SUPPORTED_THEMES = Literal[
    'abcdef',
    'abcdefDarkStyle',
    'abyss',
    'abyssDarkStyle',
    'androidstudio',
    'androidstudioDarkStyle',
    'andromeda',
    'andromedaDarkStyle',
    'atomone',
    'atomoneDarkStyle',
    'aura',
    'auraDarkStyle',
    'basicDark',
    'basicDarkStyle',
    'basicLight',
    'basicLightStyle',
    'bbedit',
    'bbeditLightStyle',
    'bespin',
    'bespinDarkStyle',
    'consoleDark',
    'consoleLight',
    'copilot',
    'copilotDarkStyle',
    'darcula',
    'darculaDarkStyle',
    'douToneLightStyle',
    'dracula',
    'draculaDarkStyle',
    'duotoneDark',
    'duotoneDarkStyle',
    'duotoneLight',
    'eclipse',
    'eclipseLightStyle',
    'githubDark',
    'githubDarkStyle',
    'githubLight',
    'githubLightStyle',
    'gruvboxDark',
    'gruvboxDarkStyle',
    'gruvboxLight',
    'kimbie',
    'kimbieDarkStyle',
    'material',
    'materialDark',
    'materialDarkStyle',
    'materialLight',
    'materialLightStyle',
    'monokai',
    'monokaiDarkStyle',
    'monokaiDimmed',
    'monokaiDimmedDarkStyle',
    'noctisLilac',
    'noctisLilacLightStyle',
    'nord',
    'nordDarkStyle',
    'okaidia',
    'okaidiaDarkStyle',
    'oneDark',
    'quietlight',
    'quietlightStyle',
    'red',
    'redDarkStyle',
    'solarizedDark',
    'solarizedDarkStyle',
    'solarizedLight',
    'solarizedLightStyle',
    'sublime',
    'sublimeDarkStyle',
    'tokyoNight',
    'tokyoNightDay',
    'tokyoNightDayStyle',
    'tokyoNightStorm',
    'tokyoNightStormStyle',
    'tokyoNightStyle',
    'tomorrowNightBlue',
    'tomorrowNightBlueStyle',
    'vscodeDark',
    'vscodeDarkStyle',
    'vscodeLight',
    'vscodeLightStyle',
    'whiteDark',
    'whiteDarkStyle',
    'whiteLight',
    'whiteLightStyle',
    'xcodeDark',
    'xcodeDarkStyle',
    'xcodeLight',
    'xcodeLightStyle',
]


class CodeMirror(ValueElement, DisableableElement, component='codemirror.js', default_classes='nicegui-codemirror'):
    VALUE_PROP = 'value'
    LOOPBACK = None

    def __init__(
        self,
        value: str = '',
        *,
        on_change: Optional[Handler[ValueChangeEventArguments]] = None,
        language: Optional[SUPPORTED_LANGUAGES] = None,
        theme: SUPPORTED_THEMES = 'basicLight',
        indent: str = ' ' * 4,
        line_wrapping: bool = False,
        highlight_whitespace: bool = False,
    ) -> None:
        """CodeMirror

        An element to create a code editor using `CodeMirror <https://codemirror.net/>`_.

        It supports syntax highlighting for over 140 languages, more than 30 themes, line numbers, code folding, (limited) auto-completion, and more.

        Supported languages and themes:
            - Languages: A list of supported languages can be found in the `@codemirror/language-data <https://github.com/codemirror/language-data/blob/main/src/language-data.ts>`_ package.
            - Themes: A list can be found in the `@uiw/codemirror-themes-all <https://github.com/uiwjs/react-codemirror/tree/master/themes/all>`_ package.

        At runtime, the methods `supported_languages` and `supported_themes` can be used to get supported languages and themes.

        :param value: initial value of the editor (default: "")
        :param on_change: callback to be executed when the value changes (default: `None`)
        :param language: initial language of the editor (case-insensitive, default: `None`)
        :param theme: initial theme of the editor (default: "basicLight")
        :param indent: string to use for indentation (any string consisting entirely of the same whitespace character, default: "    ")
        :param line_wrapping: whether to wrap lines (default: `False`)
        :param highlight_whitespace: whether to highlight whitespace (default: `False`)
        """
        super().__init__(value=value, on_value_change=self._update_cumulative)
        self._cumulative_corresponds_to_string = value
        self._cumulative_js_length: List[int] = []
        self._update_cumulative(forced=True)
        if on_change is not None:
            super().on_value_change(on_change)
        self.add_resource(Path(__file__).parent / 'lib' / 'codemirror')

        self._props['language'] = language
        self._props['theme'] = theme
        self._props['indent'] = indent
        self._props['lineWrapping'] = line_wrapping
        self._props['highlightWhitespace'] = highlight_whitespace
        self._update_method = 'setEditorValueFromProps'

    @property
    def theme(self) -> str:
        """The current theme of the editor."""
        return self._props['theme']

    @theme.setter
    def theme(self, theme: str) -> None:
        self._props['theme'] = theme
        self.update()

    def set_theme(self, theme: str) -> None:
        """Sets the theme of the editor."""
        self._props['theme'] = theme
        self.update()

    @property
    def supported_themes(self) -> List[str]:
        """List of supported themes."""
        return list(get_args(SUPPORTED_THEMES))

    @property
    def language(self) -> str:
        """The current language of the editor."""
        return self._props['language']

    @language.setter
    def language(self, language: Optional[str]) -> None:
        self._props['language'] = language
        self.update()

    def set_language(self, language: Optional[str]) -> None:
        """Sets the language of the editor (case-insensitive)."""
        self._props['language'] = language
        self.update()

    @property
    def supported_languages(self) -> List[str]:
        """List of supported languages."""
        return list(get_args(SUPPORTED_LANGUAGES))

    def _event_args_to_value(self, e: GenericEventArguments) -> str:
        """The event contains a change set which is applied to the current value."""
        return self._apply_change_set(e.args['sections'], e.args['inserted'])

    def _update_cumulative(self, *, forced=False) -> None:
        if forced or self._cumulative_corresponds_to_string != self.value:
            self._cumulative_js_length = get_cumulative_js_length(self.value)
            self._cumulative_corresponds_to_string = self.value

    def _apply_change_set(self, sections: List[int], inserted: List[List[str]]) -> str:
        # based on https://github.com/codemirror/state/blob/main/src/change.ts
        doc = self.value or ''

        def _find_python_index(js_index: int) -> int:
            return find_python_index(js_index, self._cumulative_js_length)
        assert sum(sections[::2]) == get_total_js_length(
            self._cumulative_js_length), 'Cannot apply change set to document due to length mismatch'
        pos = 0
        joined_inserts = ('\n'.join(ins) for ins in inserted)
        for section in zip_longest(sections[::2], sections[1::2], joined_inserts, fillvalue=''):
            old_len, new_len, ins = cast(Tuple[int, int, str], section)
            if new_len >= 0:
                first_index = _find_python_index(pos)
                second_index = _find_python_index(pos + old_len)
                doc = doc[:first_index] + ins + doc[second_index:]

                just_before_original_end_part = self._cumulative_js_length[second_index - 1] if second_index > 0 else 0
                original_end_part = self._cumulative_js_length[second_index:]

                self._cumulative_js_length = self._cumulative_js_length[:first_index]
                self._cumulative_js_length += [x + get_total_js_length(self._cumulative_js_length)
                                               for x in get_cumulative_js_length(ins)]
                self._cumulative_js_length += [(x - just_before_original_end_part) +
                                               get_total_js_length(self._cumulative_js_length) for x in original_end_part]
            pos += old_len
            self._cumulative_corresponds_to_string = doc
        return doc


def get_cumulative_js_length(doc: str) -> List[int]:
    """Returns a list, where for each index i, the value is the length of the string from 0 to i in UTF-16 (imagine js_len(doc[:i])"""
    return list(accumulate(len(c.encode('utf-16be'))//2 for c in doc)) if doc else []


def get_total_js_length(cumulative_js_length: List[int]) -> int:
    """Returns the length of the string in UTF-16 (imagine js_len(doc))"""
    return cumulative_js_length[-1] if cumulative_js_length else 0


def find_python_index(js_index: int, cumulative_js_length: List[int]) -> int:
    """Given a js_index, returns the position in cumulative_js_length (1-based)

    Note that 1-based indexing enables doc[:find_python_index(pos)] to replace doc[:pos]
    """
    if js_index == 0 or not cumulative_js_length:
        return 0
    lo1 = len(cumulative_js_length) - (get_total_js_length(cumulative_js_length) - js_index)
    hi1 = js_index
    lo2 = (js_index + 1) // 2
    hi2 = len(cumulative_js_length) - (get_total_js_length(cumulative_js_length) - js_index + 1) // 2
    lo = max(lo1, lo2, 0)
    hi = min(hi1, hi2, len(cumulative_js_length))
    return bisect.bisect_right(cumulative_js_length, js_index, lo, hi)

File: nicegui/elements/codemirror.js
export default {
  template: `
    <div></div>
  `,
  props: {
    value: String,
    language: String,
    theme: String,
    resource_path: String,
    lineWrapping: Boolean,
    minHeight: String,
    fixedHeight: String,
    maxHeight: String,
    disable: Boolean,
    indent: String,
    highlightWhitespace: Boolean,
  },
  watch: {
    value(new_value) {
      this.set_editor_value(new_value);
    },
    language(new_language) {
      this.set_language(new_language);
    },
    theme(new_theme) {
      this.set_theme(new_theme);
    },
    disable(new_disable) {
      this.set_disabled(new_disable);
    },
  },
  data() {
    return {
      editorPromise: new Promise((resolve) => {
        this.resolveEditor = resolve;
      }),
    };
  },
  methods: {
    // Find the language's extension by its name. Case insensitive.
    find_language(name) {
      for (const language of this.languages)
        for (const alias of [language.name, ...language.alias])
          if (name.toLowerCase() === alias.toLowerCase()) return language;

      console.error(`Language not found: ${this.language}`);
      console.info("Supported languages names:", languages.map((lang) => lang.name).join(", "));
      return null;
    },
    // Get the names of all supported languages
    async get_languages() {
      if (!this.editor) await this.editorPromise;
      // Over 100 supported languages: https://github.com/codemirror/language-data/blob/main/src/language-data.ts
      return this.languages.map((lang) => lang.name);
    },
    set_language(language) {
      const lang_description = this.find_language(language, this.languages);
      if (!lang_description) {
        console.error("Language not found:", language);
        return;
      }

      lang_description.load().then((extension) => {
        this.editor.dispatch({
          effects: this.languageConfig.reconfigure([extension]),
        });
      });
    },
    async get_themes() {
      if (!this.editor) await this.editorPromise;
      // `this.themes` also contains some non-theme objects
      // The real themes are Arrays
      return Object.keys(this.themes).filter((key) => Array.isArray(this.themes[key]));
    },
    set_theme(theme) {
      const new_theme = this.themes[theme];
      if (new_theme === undefined) {
        console.error("Theme not found:", theme);
        return;
      }
      this.editor.dispatch({
        effects: this.themeConfig.reconfigure([new_theme]),
      });
    },
    set_editor_value(value) {
      if (!this.editor) return;
      if (this.editor.state.doc.toString() === value) return;

      this.emitting = false;
      this.editor.dispatch({ changes: { from: 0, to: this.editor.state.doc.length, insert: value } });
      this.emitting = true;
    },
    set_disabled(disabled) {
      this.editor.dispatch({
        effects: this.editableConfig.reconfigure(this.editableStates[!disabled]),
      });
    },
  },
  async mounted() {
    const {
      EditorView,
      Compartment,
      basicSetup,
      keymap,
      indentWithTab,
      languages,
      themes,
      oneDark,
      indentUnit,
      highlightWhitespace,
    } = await import(`${this.resource_path}/editor.js`);

    // This flag is used to prevent emitting the same value
    // again when the editor was just updated from the server
    this.emitting = true;

    const changeListener = EditorView.updateListener.of((update) => {
      if (!this.emitting) return;
      if (!update.docChanged) return;
      const value = update.state.doc.toString();
      this.$emit("update:value", value);
    });

    // The Compartments are used to change the properties of the editor ("extensions") dynamically
    this.themes = { ...themes, oneDark: oneDark };
    this.themeConfig = new Compartment();
    this.languages = languages;
    this.languageConfig = new Compartment();
    this.editableConfig = new Compartment();
    this.editableStates = { true: EditorView.editable.of(true), false: EditorView.editable.of(false) };

    const extensions = [
      basicSetup,
      changeListener,
      // Enables the Tab key to indent the current lines https://codemirror.net/examples/tab/
      keymap.of([indentWithTab]),
      // Sets indentation https://codemirror.net/docs/ref/#language.indentUnit
      indentUnit.of(this.indent),
      // We will set these compartments later and dynamically through props
      this.themeConfig.of([]),
      this.languageConfig.of([]),
      this.editableConfig.of([]),
    ];

    if (this.lineWrapping) extensions.push(EditorView.lineWrapping);
    if (this.highlightWhitespace) extensions.push([highlightWhitespace()]);

    // Convenience function to add theme properties below
    const addToTheme = (content) => {
      extensions.push(EditorView.theme(content));
    };

    // Setting the height properly through tailwind is likely not possible,
    // so we use the recommended ways from https://codemirror.net/examples/styling/
    if (this.fixedHeight) {
      addToTheme({
        "&": { height: this.fixedHeight },
        ".cm-scroller": { overflow: "auto" },
      });
    } else {
      if (this.maxHeight)
        addToTheme({
          "&": { "max-height": this.maxHeight },
          ".cm-scroller": { overflow: "auto" },
        });
      if (this.minHeight)
        addToTheme({
          ".cm-content, .cm-gutter": { minHeight: this.minHeight },
        });
    }

    this.editor = new EditorView({
      doc: this.value,
      extensions: extensions,
      parent: this.$el,
    });

    this.resolveEditor(this.editor);

    this.set_language(this.language);
    this.set_theme(this.theme);
    this.set_disabled(this.disable);
  },
};
